:book: Add the quickstart details for KubeVirt

nunnatsa opened the pull request:
Signed-off-by: Nahshon Unna-Tsameret <nunnatsa@redhat.com>

**What this PR does / why we need it**:
This PR add many details for the quickstart guide, to explain how to deploy a guest cluster using the KubeVirt provider.

nunnatsa commented:
/cc @davidvossel

sbueringer commented on docs/book/src/user/quick-start.md:
```diff
@@ -1146,6 +1223,92 @@ kubectl --kubeconfig=./capi-quickstart.kubeconfig get nodes
 Calico not required for vcluster.
 
+{{#/tab }}
+{{#tab kubevirt}}
```
```suggestion
{{#tab Kubevirt}}
```
(same in l.1204, 1041 and 1077)

nunnatsa replied:
Changed all to `KubeVirt`

sbueringer commented on docs/book/src/user/quick-start.md:
```diff
@@ -436,7 +436,65 @@ clusterctl init --infrastructure ibmcloud
+configurations to our kind cluster.
+
+##### Install the calico CNI
+The default kind's CNI (kindnet) will not work for the guest cluster. We'll use calico CNI. Download the calico manifest and use it to create the required resources, as described [here](https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises#install-calico); e.g.:
```
```suggestion
kind's default CNI (kindnet) will not work for the guest cluster. We'll use calico CNI. Download the calico manifest and use it to create the required resources, as described [here](https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises#install-calico); e.g.:
```

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -436,7 +436,65 @@ clusterctl init --infrastructure ibmcloud
+The default kind's CNI (kindnet) will not work for the guest cluster. We'll use calico CNI. Download the calico manifest and use it to create the required resources, as described [here](https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises#install-calico); e.g.:
+```bash
+curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.3/manifests/calico.yaml -O
+kubectl create -f calico.yaml
````
Does this mean that after that kindnet & Calico are installed? Isn't that a problem?

nunnatsa replied:
> Does this mean that after that kindnet & Calico are installed? Isn't that a problem?

This is a good question, @sbueringer. So in my environment it's working (tested over and over again). It is possible to disable kindnet (https://kind.sigs.k8s.io/docs/user/configuration/#disable-default-cni) but I thought it will make the document a bit too complicated, because I'll need a custom kind cluster creation for KubeVirt, and since it does work without this modification, I thought it will be simpler to leave it as is. WDYT?

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -1015,6 +1073,26 @@ clusterctl generate cluster ${CLUSTER_NAME} \
+```
+Now, modify the service to be with the`LoadBalancer` type:
+```bash
+sed -i 's/type: ClusterIP/type: LoadBalancer/' cluster-template.yaml
````
Would it be possible to do this change to cluster-template in kubevirt itself?

Would be easier for users vs. telling them to modify the released cluster-template

nunnatsa replied:
@davidvossel - what do you think?

nunnatsa replied:
OK - working on it, here: https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/202

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -1146,6 +1223,92 @@ kubectl --kubeconfig=./capi-quickstart.kubeconfig get nodes
+```
+
+We'll need to prevent conflicts with the kind cluster network, buy modifying the default calico settings:
+* change the CIDR to a non-conflict range
+* Change the value of the `CLUSTER_TYPE` environment variable to `k8s`
+* Change the value of the `CALICO_IPV4POOL_IPIP` environment variable to `Never`
+* Change the value of the `CALICO_IPV4POOL_VXLAN` environment variable to `Always`
+* add the `FELIX_VXLANPORT` environment variable with the value of a non-conflict port, e.g. `"6789"`.
````
```suggestion
* Change the CIDR to a non-conflict range
* Change the value of the `CLUSTER_TYPE` environment variable to `k8s`
* Change the value of the `CALICO_IPV4POOL_IPIP` environment variable to `Never`
* Change the value of the `CALICO_IPV4POOL_VXLAN` environment variable to `Always`
* Add the `FELIX_VXLANPORT` environment variable with the value of a non-conflicting port, e.g. `"6789"`.
```
nit

sbueringer commented on docs/book/src/user/quick-start.md:
```diff
@@ -436,7 +436,65 @@ clusterctl init --infrastructure ibmcloud
+Please visit the [Kubevirt project][Kubevirt provider] for more information.
+
+KubeVirt is a cloud native virtualization solution. The meaning for that for our guest cluster, is that the virtual
+machines will be created within the kind cluster, as a kubernetes resources. To make it work, we'll need to add some
```
```suggestion
machines will be created within the kind cluster, as a Kubernetes resources. To make it work, we'll need to add some
```

sbueringer commented on docs/book/src/user/quick-start.md:
```diff
@@ -436,7 +436,65 @@ clusterctl init --infrastructure ibmcloud
+machines will be created within the kind cluster, as a kubernetes resources. To make it work, we'll need to add some
+configurations to our kind cluster.
+
+##### Install the calico CNI
```
```suggestion
##### Install the Calico CNI
```

sbueringer replied:
(same in other places)

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -436,7 +436,65 @@ clusterctl init --infrastructure ibmcloud
+kubectl create -f calico.yaml
+```
+
+##### Install metallb for load balancing
````
```suggestion
##### Install MetalLB for load balancing
```

sbueringer commented on docs/book/src/user/quick-start.md:
```diff
@@ -795,13 +853,13 @@ Please visit the [IBM Cloud provider] for more information.
+export CAPK_GUEST_K8S_VERSION="${CAPK_GUEST_K8S_VERSION:-v1.23.10}"
+export CRI_PATH="${CRI_PATH:-/var/run/containerd/containerd.sock}"
+export NODE_VM_IMAGE_TEMPLATE="${NODE_VM_IMAGE_TEMPLATE:-quay.io/capk/ubuntu-2004-container-disk:${CAPK_GUEST_K8S_VERSION}}"
+export IMAGE_REPO="${IMAGE_REPO:-k8s.gcr.io}"
```
fyi. The new default with v1.25 is registry.k8s.io (there's a redirect right now, so both serve the same images)

Please consider using the new registry

nunnatsa replied:
The latest image is for 1.23

@davidvossel - are we planing to introduce an image for 1.25?

nunnatsa replied:
For some reason this is just not working and I can't debug it. I will leave the original value of `k8s.gcr.io`.

sbueringer replied:
Okay, please follow-up on this. Kubernetes is literally running out of money with the old registry. So we're trying everything we can going forward to use the new registry as it's cheaper.

sbueringer commented on docs/book/src/user/quick-start.md:
```diff
@@ -1015,6 +1073,26 @@ clusterctl generate cluster ${CLUSTER_NAME} \
+{{#/tab }}
+{{#tab kubevirt}}
+
+First, download the cluster manifest, because we'll need to modify it a bit:
```
```suggestion
First, download the cluster template, because we'll need to modify it a bit:
```

sbueringer commented on docs/book/src/user/quick-start.md:
```diff
@@ -1117,14 +1195,13 @@ Note: To use the default clusterctl method to retrieve kubeconfig for a workload
 
 </aside>
 
-{{#/tab }}
```
Why was this dropped?

nunnatsa replied:
ooops

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -1146,6 +1223,92 @@ kubectl --kubeconfig=./capi-quickstart.kubeconfig get nodes
+capi-quickstart-md-0-zht5j            164m   Running   10.244.82.17   kind-control-plane   True
+```
+
+We'll need to prevent conflicts with the kind cluster network, buy modifying the default calico settings:
````
```suggestion
We'll need to prevent conflicts with the kind cluster network, by modifying the default calico settings:
```

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -1146,6 +1223,92 @@ kubectl --kubeconfig=./capi-quickstart.kubeconfig get nodes
+```
+
+We'll need to prevent conflicts with the kind cluster network, buy modifying the default calico settings:
+* change the CIDR to a non-conflict range
````
```suggestion
* change the CIDR to a non-conflicting range
```

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -1146,6 +1223,92 @@ kubectl --kubeconfig=./capi-quickstart.kubeconfig get nodes
+'/# Set Felix endpoint to host default action to ACCEPT./a\            - name: FELIX_VXLANPORT\n              value: "6789"' \
+calico.yaml
+```
+Now, deploy the calico CNI on the guest cluster:
````
```suggestion
Now, deploy the calico CNI on the workload cluster:
```
Should we call it workload cluster? (given that's how CAPI calls it in general)

(also in other places)

sbueringer reviewed (commented):
A few nits.

Thank you very much for adding more documentation!

davidvossel reviewed (approved):
/lgtm

I went through all these instructions and it worked perfect for me locally.

davidvossel commented:
@sbueringer would you be able to take another look at this? I ran through the steps locally and it works for me. Now, we need an approve. Thanks!

enxebre commented:
Thanks!
/lgtm
for Stefan to have the final look.

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -1015,6 +1105,23 @@ clusterctl generate cluster ${CLUSTER_NAME} \
+
+Create the workload cluster using the load balancer template  
+```bash
+CAPK_VER=$(curl "https://api.github.com/repos/kubernetes-sigs/cluster-api-provider-kubevirt/releases/latest" | jq -r ."tag_name")
+
+curl -L -O "https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/releases/download/${CAPK_VER}/cluster-template-lb.yaml"
+
+clusterctl generate cluster capi-quickstart \
+  --kubernetes-version ${CAPK_GUEST_K8S_VERSION} \
+  --control-plane-machine-count=1 \
+  --worker-machine-count=1 \
+  --from cluster-template-lb.yaml \
+  > capi-quickstart.yaml
````
```suggestion
clusterctl generate cluster capi-quickstart --flavor lb \
  --kubernetes-version ${CAPK_GUEST_K8S_VERSION} \
  --control-plane-machine-count=1 \
  --worker-machine-count=1 \
  --from cluster-template-lb.yaml \
  > capi-quickstart.yaml
```

Should work now

nunnatsa replied:
:+1: :+1: :+1: 

These are good news. I opened a related issue (#7455) + PR to fix it (#7453), in order to use the `--form` flag to get a template from release assets. I still think it should be merged, but in this case, the `--flavor` flag will work.

sbueringer replied:
Sounds good to me!

nunnatsa replied:
@sbueringer - done.

sbueringer replied:
Thx!

sbueringer reviewed (commented):
last nit from my side

cc @fabriziopandini

sbueringer commented:
/lgtm

I would like a review from Fabrizio as well

/assign @fabriziopandini

fabriziopandini commented on docs/book/src/user/quick-start.md:
```diff
@@ -93,6 +93,45 @@ a target [management cluster] on the selected [infrastructure provider].
+   {{#tab KubeVirt}}
+
+   #### Create the Kind Cluster
+   To allow using LoadBalancer, we can't use the kind's default CNI (kindnet), but we'll need to install another CNI, like Calico. In order to do that, we'll need first to initiate the kind cluster with two modifications:
```
I think it could be useful to add a note explaining why it is required to have a load balancer in the management cluster (which is something we do not have for other providers)?

fabriziopandini commented on docs/book/src/user/quick-start.md:
```diff
@@ -434,9 +473,60 @@ clusterctl init --infrastructure ibmcloud
+within KubeVirt virtual machines. In order to make this work, we need to apply some configurations to our kind cluster.
+
+#### Install MetalLB for load balancing
+We'll need to enable load-balance services in our kind cluster. For that, we'll install MetalLB, as describe [here](https://metallb.universe.tf/installation/#installation-by-manifest); e.g.
```
nit
```suggestion
We'll need to enable load-balance services in our kind cluster. For this tutorial, we'll install MetalLB, as described [here](https://metallb.universe.tf/installation/#installation-by-manifest); e.g.
```
Assuming that any LB could work

fabriziopandini commented on docs/book/src/user/quick-start.md:
````diff
@@ -1015,6 +1105,20 @@ clusterctl generate cluster ${CLUSTER_NAME} \
+```bash
+clusterctl generate cluster capi-quickstart \
+  --infrastructure="kubevirt" \
+  --flavor lb \
````
is there a reason why we are not using the default template?

fabriziopandini commented on docs/book/src/user/quick-start.md:
```diff
@@ -793,15 +883,15 @@ export IBMPOWERVS_NETWORK_NAME=<your-capi-network-name>
+export CAPK_GUEST_K8S_VERSION="v1.23.10"
+export CRI_PATH="/var/run/containerd/containerd.sock"
+export NODE_VM_IMAGE_TEMPLATE="quay.io/capk/ubuntu-2004-container-disk:${CAPK_GUEST_K8S_VERSION}"
+export IMAGE_REPO="k8s.gcr.io"
```
can we switch this to the new registry, images should be already available in both (to be verified)

nunnatsa replied:
I tried that several times and it's just not working. After looking at the code, I think it should only work from K8s v1.25+.

I also noticed that other providers just dropped this setting, so I opened a PR to drop it also in Kubevirt provider (kubernetes-sigs/cluster-api-provider-kubevirt#207) and I hope it will be included in the next version.

sbueringer replied:
We are using the new registry for Kubernetes >= 1.22

sbueringer replied:
+1 to your PR. I would definitely recommend not exposing IMAGE_REPO as an env var

fabriziopandini commented on docs/book/src/user/quick-start.md:
````diff
@@ -1146,6 +1250,111 @@ kubectl --kubeconfig=./capi-quickstart.kubeconfig get nodes
+capi-quickstart-md-0-zht5j            164m   Running   10.244.82.17   kind-control-plane   True
+```
+
+We'll need to prevent conflicts with the kind cluster network, by modifying the default Calico settings:
````
I think it will be beneficial for users to explain why we need this change

fabriziopandini reviewed (commented):
overall lgtm
long term I would like to move the part about docker hub rate limit in the troubleshooting guide, but this can be a follow up PR

CecileRobertMichon commented on docs/book/src/user/quick-start.md:
````diff
@@ -93,6 +93,45 @@ a target [management cluster] on the selected [infrastructure provider].
+   #### Install the Calico CNI
+   Download the Calico manifest and use it to create the required resources, as described [here](https://projectcalico.docs.tigera.io/getting-started/kubernetes/self-managed-onprem/onpremises#install-calico); e.g.:
+   ```bash
+   curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.3/manifests/calico.yaml -O
````
Do we want to pin Calico to a specific version? It will require frequent updates to the book in order to not get outdated which might make this doc harder to maintain

sbueringer replied:
In the "others..." tab it has been proven to be sustainable as we usually bump calico as well when we bump Kubernetes.

I don't know how sustainable it is for kubevirt (as we wouldn't bump that when we bump for others as we don't have the time to test it).

We pin in the "others..." tab to ensure we have something which has been tested manually at least once. We couldn't guarantee this with some sort of "latest".

nunnatsa replied:
Thanks for your comment @CecileRobertMichon. I was wandering about it when I wrote it. But this is not only a matter of versions, because new calico version my also bring relevant API changes. So I prefer to pin it for now, and monitor it. Any change will require some manual testing (i.e. go over the whole guide again and again to make sure everything is still working).

CecileRobertMichon replied:
That's fair 👍 SGTM

CecileRobertMichon commented on docs/book/src/user/quick-start.md:
````diff
@@ -1146,6 +1250,111 @@ kubectl --kubeconfig=./capi-quickstart.kubeconfig get nodes
+
+The following script downloads the Calico manifest and modifies the required field. The CIDR and the port values are examples.
+```bash
+curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.3/manifests/calico.yaml -o calico-workload.yaml
````
same question here about calico version

davidvossel reviewed (commented):
/lgtm

g-gaston commented on docs/book/src/user/quick-start.md:
````diff
@@ -93,6 +93,56 @@ a target [management cluster] on the selected [infrastructure provider].
+
+   Download the Calico manifest and use it to create the required resources; e.g.:
+   ```bash
+   curl https://raw.githubusercontent.com/projectcalico/calico/v3.24.4/manifests/calico.yaml -O
+   kubectl create -f calico.yaml
````
nit: you could make this easier
```suggestion
   kubectl create -f https://raw.githubusercontent.com/projectcalico/calico/v3.24.4/manifests/calico.yaml
```

g-gaston commented on docs/book/src/user/quick-start.md:
```diff
@@ -93,6 +93,56 @@ a target [management cluster] on the selected [infrastructure provider].
+   To allow using a LoadBalancer service, we can't use the kind's default CNI (kindnet), but we'll need to install
+   another CNI, like Calico. In order to do that, we'll need first to initiate the kind cluster with two modifications:
+   1. Disable the default CNI
+   2. Add the docker credentials to the cluster, to avoid the docker hub pull rate limit.
```
How frequently does this happen? Are the kubevirt images in docker hub?
Asking because if it's not always necessary, we can omit it and maybe add it as troubleshooting step/advice. It adds one more thing the user needs to do and since this is a quick start, the less steps, the better.

nunnatsa replied:
@g-gaston, this is pretty frequent. I would leave it here because (a) we need the configuration file anyway to disable kindnet CNI, and (b) because I think it's more than a regular troubleshoot, because starting the kind cluster is a basic pre-requisite. To re-do it, you must start over the whole process. So I think we need the right setting in advance.

g-gaston commented on docs/book/src/user/quick-start.md:
```diff
@@ -93,6 +93,56 @@ a target [management cluster] on the selected [infrastructure provider].
+   - role: control-plane
+     extraMounts:
+      - containerPath: /var/lib/kubelet/config.json
+        hostPath: <YOUR DOCKER CONFIG FILE PATH>
```
If you want to keep this because the rate limit happens frequently, I would suggest to add a bit more info on how to create/find this file, or maybe just link to the official docs if there is any.

One could argue this is pretty basic and any CAPI user should already know how to do it. However, this is a quick start guide, so the lower the barrier of entry, the better :)

g-gaston reviewed (commented):
lgtm

nunnatsa commented:
@fabriziopandini, @sbueringer, @davidvossel - I made a few changes:

1. After releasing a new CAPK version, we can now drop the image registry environment variable, so I removed it.
2. Moved the workload cluster docker hub rate limit handling to a troubleshooting section.
3. Add a few more explanations.

Could you please review again?

sbueringer commented on docs/book/src/user/quick-start.md:
````diff
@@ -442,11 +493,61 @@ clusterctl init --infrastructure kubekey
+kubectl wait pods -n metallb-system -l app=metallb,component=speaker --for=condition=Ready --timeout=2m
+```
+
+Now, we'll Create the `IPAddressPool` and the `L2Advertisement` custom resources. The script below creates the CRs with
````
```suggestion
Now, we'll create the `IPAddressPool` and the `L2Advertisement` custom resources. The script below creates the CRs with
```
nit

nunnatsa replied:
Thanks @sbueringer - done

sbueringer replied:
Thx!

sbueringer commented:
@nunnatsa Just one last nit from my side.

Very very nice :) Happy to see this land.

Appreciate your patience in addressing the review comments!!

sbueringer commented:
/cherry-pick release-1.3

sbueringer commented:
/cherry-pick release-1.2

sbueringer commented:
/lgtm

/assign @fabriziopandini

g-gaston reviewed (approved):
lgtm

fabriziopandini commented:
Great Job!
/lgtm
/approve
